### PR TITLE
Variable moduli for elasticity tensor

### PR DESCRIPTION
**Pull request by bwspenc**
This implements a material that computes elastic moduli that depend on other materials.  I'm following this approach instead of my previous one that used functions and was specific to temperatures per feedback from @dschwen.

Because we have piecewise linear functions that describe the dependence of these moduli on temperature, I also implemented a general Material that incorporates a LinearInterpolation object to define the moduli within the Material, rather than using a Function.

@WilkAndy This is the code we discussed yesterday.

### CODE REVIEW

File: framework/src/materials/PiecewiseLinearInterpolationMaterial.C [hunk 1]
```diff
@@ -0,0 +1,87 @@
+/****************************************************************/
```
Comment (bwspenc) [inline framework/src/materials/PiecewiseLinearInterpolationMaterial.C]: Any thoughts on the name for this?  @permcody? There are a couple of other materials of this nature in this directory with names that start with 'Generic', but I didn't include that to keep the name shorter.

**Comment by WilkAndy** [comment]
Thanks for this, @bwspenc .   I hope we can do some heterogeneous rock modelling with it.

File: framework/src/materials/PiecewiseLinearInterpolationMaterial.C (hunk 1 above)
Comment (permcody) [inline framework/src/materials/PiecewiseLinearInterpolationMaterial.C]: This is fine. If others are in favor of dropping generic from the existing similar objects. I'm also fine with that.

**Comment by permcody** [comment]
Ping @dschwen

File: framework/include/materials/PiecewiseLinearInterpolationMaterial.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+public:
+  PiecewiseLinearInterpolationMaterial(const InputParameters & parameters);
+
+protected:
+  virtual void computeQpProperties() override;
+
+  // Name of the property to be computed
+  std::string _prop_name;
+
+  // Value of the coupled variable to be used as the abscissa in the piecewise linear interpolation
```
Comment (dschwen) [inline framework/include/materials/PiecewiseLinearInterpolationMaterial.h]: make both `///`

File: framework/src/materials/PiecewiseLinearInterpolationMaterial.C [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  {
+    std::vector<Real> xy = getParam<std::vector<Real> >("xy_data");
+    unsigned int xy_size = xy.size();
+    if (xy_size % 2 != 0)
+      mooseError("In PiecewiseLinearInterpolationMaterial ", _name, ": Length of data provided in 'xy_data' must be a multiple of 2.");
+
+    unsigned int x_size = xy_size/2;
+    x.reserve(x_size);
+    y.reserve(x_size);
+    for (unsigned int i=0; i<xy_size/2; ++i)
```
Comment (dschwen) [inline framework/src/materials/PiecewiseLinearInterpolationMaterial.C]: operator spacing

**Review by dschwen — CHANGES_REQUESTED** [review]
Looks good otherwise

File: modules/combined/tests/thermal_elastic/thermal_elastic.i [hunk 4]
```diff
@@ -83,133 +85,36 @@
...
-  [./solid]
-    disp_x = disp_x
-    disp_y = disp_y
-    disp_z = disp_z
+[Modules/TensorMechanics/Master]
+  [./all]
+    add_variables = true
+    generate_output = 'stress_xx stress_yy stress_zz stress_xy stress_xz stress_yz'
+    strain = FINITE
+    block = '1 2 3 4 5 6 7'
```
Comment (dschwen) [inline modules/combined/tests/thermal_elastic/thermal_elastic.i]: if you omit the block parameter no restriction is applied (which is what the `all` name seems to indicate you want)

**Comment by bwspenc** [comment]
@dschwen I addressed the issues you pointed out.

**Review by dschwen — APPROVED** [review]

**Comment by bwspenc** [comment]
@permcody The documentation build is failing for an unrelated reason (recent deprecation of StressDivergencePFFracTensors in another commit?).  Do we need that to build cleanly before we merge this, or has someone already addressed this problem?  Since @dschwen isn't available today, can you help me get this merged?

**Comment by permcody** [comment]
As far as process goes, we do not require a clean build in order to merge. That's just too restrictive. All you *should* do is describe why you are merging with a broken build for auditing purposes.

**Comment by permcody** [comment]
Only the Documentation target is breaking for unrelated reasons. Merging...

**Pull request merged**